src/particles/PhysicsParticles.hpp

```diff
@@ -0,0 +1,365 @@
+							    amrex::GpuArray<amrex::Real, AMREX_SPACEDIM> const &dxi) const noexcept
+	{
+		amrex::ParticleInterpolator::Linear interp(p, plo, dxi);
+		interp.ParticleToMesh(p, radEnergySource, start_part_comp, start_mesh_comp, num_comp,
+				      [this, dxi] AMREX_GPU_DEVICE(const ParticleType &part, int comp) {
+					      if (current_time < part.rdata(birthTimeIndex) || current_time >= part.rdata(birthTimeIndex + 1)) {
+						      return 0.0;
+					      }
+					      return part.rdata(comp) * (AMREX_D_TERM(dxi[0], *dxi[1], *dxi[2]));
+				      });
+	}
+};
+
+// Base class for physics particle descriptors
+class PhysicsParticleDescriptor
+{
+      protected:
+	int massIndex_{-1};						   // index for gravity mass, -1 if not used
+	int lumIndex_{-1};						   // index for radiation luminosity, -1 if not used
+	int birthTimeIndex_{-1};					   // index for birth time, -1 if not used
+	bool interactsWithHydro_{false};				   // whether particles interact with hydro
+	amrex::ParticleContainerBase *neighborParticleContainer_{nullptr}; // non-owning pointer to particle container
+
+      public:
+	PhysicsParticleDescriptor(int mass_idx, int lum_idx, int birth_time_idx, bool hydro_interact)
+	    : massIndex_(mass_idx), lumIndex_(lum_idx), birthTimeIndex_(birth_time_idx), interactsWithHydro_(hydro_interact)
+	{
+	}
+	virtual ~PhysicsParticleDescriptor() = default;
+
+	// Getters
+	[[nodiscard]] auto getMassIndex() const -> int { return massIndex_; }
+	[[nodiscard]] auto getLumIndex() const -> int { return lumIndex_; }
+	[[nodiscard]] auto getBirthTimeIndex() const -> int { return birthTimeIndex_; }
+	[[nodiscard]] auto getInteractsWithHydro() const -> bool { return interactsWithHydro_; }
+
+	virtual void hydroInteract() {} // Default no-op
+
+	// Delete copy/move constructors/assignments
+	PhysicsParticleDescriptor(const PhysicsParticleDescriptor &) = delete;
+	PhysicsParticleDescriptor &operator=(const PhysicsParticleDescriptor &) = delete;
+	PhysicsParticleDescriptor(PhysicsParticleDescriptor &&) = delete;
+	PhysicsParticleDescriptor &operator=(PhysicsParticleDescriptor &&) = delete;
+
+	// Setter for particle container
+	template <typename ParticleContainerType> void setParticleContainer(ParticleContainerType *container) { neighborParticleContainer_ = container; }
+
+	// Virtual operations that will be called on the particle container
+	virtual void redistribute(int lev) = 0;
+	virtual void redistribute(int lev, int ngrow) = 0;
+	virtual void writePlotFile(const std::string &plotfilename, const std::string &name) = 0;
+	virtual void writeCheckpoint(const std::string &checkpointname, const std::string &name, bool include_header) = 0;
+	virtual void depositRadiation(amrex::MultiFab &radEnergySource, int lev, amrex::Real current_time, int lumIndex, int birthTimeIndex, int nGroups) = 0;
+	virtual void depositMass(amrex::Vector<amrex::MultiFab> &rhs, int finest_lev, amrex::Real Gconst, int massIndex) = 0;
```

Comment:
I think `depositMass` could also be implemented here in the base class.

Reply:
Unfortrunately, no. See my comments above.